### DISCUSSION TITLE
Multiple Variables and Solvers

### DISCUSSION THREAD

**Discussion by czadeh** [idaholab/moose#21918] [Q&A General] [answered]
Hi,

I am working on a simulation where I have multiple variables.

The first variable is best solved by PJFNK. It is temperature in a heat transfer problem.

The second can be solved explicitly with Euler's method. The equation is a first order ODE coupled to temperature.
However, in my current implementation I solve it with PJFNK which is overkill for this task. It may be more appropriate to solve this as a material property or aux variable but I am not sure if that is technically possible or what the tradeoff is.

The third could be solved with Newton's method. The equation is a first order ODE coupled to temperature.
I have not implemented it yet or decided if it should be a variable, aux variable, or material property.

My main question is: is there a way to use different solvers in the same simulation (i.e., PJFNK, Euler's method, and Newton's method)?

Alternatively, is it advantageous to implement variables 2 and 3 as aux variables or material properties? Would it be possible to solve ODEs in this case?

Thank you,
Cameron

**Comment by GiudGiud** [comment]
Hello

You may use NodalKernels to write ODEs in MOOSE.

The easiest approach to have these varied solve methods would be to use the MultiApps system and have each solve in a different application. 

Variable 2 and 3 may be auxvariables or material properties if you can write down the entire solve in a single auxkernel / material. if you need to use the main nonlinear solve for it, then that wont work. 

Guillaume

**Reply by czadeh** [reply]
I'm looking into the NodalKernels system but there is not much documentation about it so it seems more difficult.

The MultiApp system could be the best solution. Could I make it so that the main app and each of the MultiApps cover the same domain but just solve independently and exchange information?

The last option of writing the entire solve in the kernel seems possible but unnecessarily difficult when the MultiApp system can be used.

Thanks

**Reply by czadeh** [reply]
Also, if I wish to use mesh adaptivity in the main app, then will it cause problems when exchanging information between the main app and sub apps?

**Reply by GiudGiud** [reply]
Documentation is poor but this is the way. please have a look at examples in the test folder. 

Even within the MultiApps system you will want to use NodalKrnels for the ODEs

Yes you can have the different apps cover the same domain. It will make things easy to transfer between apps actually

no mesh adaptivity should be naturally handled. You just wont be able to use some classes of transfers

**Reply by czadeh** [reply]
I have set up a MultiApp: main app with temperature, sub-app with my second variable. Initial testing checks out.

Now I need to convert my Kernel for the second variable into a NodalKernel. I have my Kernel return the residual of the variable involving the time derivative so I had it inherit from ADTimeDerivative.h. 

Can I use a similar approach for the NodalKernel? i.e., have it inherit from TimeDerivativeNodalKernel.h and run `Real TimeDerivativeNodalKernel::computeQpResidual()` in the source file to return the residual.

This also ties into my last question: do I need to include the `computeQpJacobian()` method as well in my NodalKernel.C because it was there in TimeDerivativeNodalKernel.C?

I want to solve using an explicit method like the ones in the timeintegrators system, so I do not need to solve the Jacobian, right?

Thanks,
Cameron

**Reply by GiudGiud** [reply]
for fully explicit time stepping you should be able to not implement the Jacobian. The 'solve' on every time step becomes very simple, no nonlinear solve anymore

**Reply by GiudGiud** [reply]
but if you want to do implicit later on, you must be careful to make sure everything is implemented, not just inherited from the base class. This could be a hard one to debug if you loose track of it. 

I would include the method and add a warning in there.

**Reply by czadeh** [reply]
I just made my first attempt at the NodalKernel for the second variable (var2). My problem now is that 

- The residual for var2 in the NodalKernel must be a Real type because I need to use `Real computeQpResidual()` 
- var2 is coupled to temperature which is an ADVariableValue
- In my previously existing Kernel for var2, I was using `ADReal precomputeQpResidual()`

Since temperature is passed through the transfer system and ends up as an AuxVariable in the sub-app for var2, is it an ADVariableValue at that point still? If not, what is the data type? If I can treat it as a Real that would solve this problem.

**Reply by GiudGiud** [reply]
AuxVariables are not AD. So it's a regular variable at this point. 

If it's retrieved as an AD variable for any reason, you can always do <number>.value() to get the Real out of the ADReal.

**Reply by czadeh** [reply]
Thank you, I got my NodalKernel to compile. I set up my new MultiApp system with it, updating the main-app to run the new sub-app and changing the sub-app to use the NodalKernel and the Executioner to use the TimeIntegrator system with ActuallyExplicitEuler. When I run it, it converges for time step 0 and time step 1, but after that I get a ton of these `printf("Jacobian warning")` messages in the terminal and it does not converge. I put that print statement inside of the `computeQpJacobian()` method along with `return 0;`

I am surprised to see those messages because I would think that ActuallyExplicitEuler would not call that method, but now I am confused.

The final error is 

```
*** ERROR ***
The following error occurred in the object "TimeStepper", of type "ConstantDT".

Solve failed and timestep already at or below dtmin, cannot continue!
```

**Reply by GiudGiud** [reply]
I'm confused too.
@friedmud @joshuahansel what are we missing here? Why is the Jacobian used for explicit time-stepping?

**Reply by joshuahansel** [reply]
You must compute the Jacobian of time kernels still - this is the mass matrix.

**Reply by joshuahansel** [reply]
So if the NodalKernel in question does not have the "time" tag, and it is computing Jacobians, that seems like a bug.

**Reply by GiudGiud** [reply]
it does, he wrote the whole ODE in the time derivative kernel. 
Well I guess now we know what's the issue with that

**Reply by czadeh** [reply]
So I guess if I want to go this route I need to write the Jacobian?

I was able to use ActuallyExplicitEuler on my regular ADTimeDerivativeKernel and still saw performance improvements without a Jacobian.

**Reply by friedmud** [reply]
Even ActuallyExplicitEuler needs the Jacobian... depending on what option you choose it may need the full mass matrix or a lumped version of it.

**Comment by joshuahansel** [comment, marked as answer]
Just curious, it seems like your question is motivated by the idea that solving everything in a single system with a single solver is "overkill" (and I suppose, not as efficient). Have you tried this already, and you're not satisfied with the performance?

To use different solvers, you'll need to use the MultiApp system as @GiudGiud suggested, but as you may realize, it does come at the cost of the full coupling of your 3 variables (you'll need to either loosely couple or tightly couple through Picard iterations). The question of which is ultimately the most efficient approach will depend on your system and size, and I don't have enough experience to say what the most efficient approach is, but it's a consideration to make.

**Reply by czadeh** [reply]
You raise a very good point. I have compared them and seen some modest improvement in speed. My most recent test was with a MultiApp:

- 1 variable in the main
- 1 variable in the sub
- 10 x 10 x 10 nodes
- 100 time steps
- 3 material properties
- Using my ADTimeDerivative Kernel for the variable in the sub app

And the difference was one was solved with PJFNK in the sub app while the other was solved with ActuallyExplicitEuler in the sub app.
I saw a time of 21.40 s for PJFNK and 17.52 s for ActuallyExplicitEuler. The global truncation error was around 3% for Euler's method compared to PJFNK after 100 iterations.

Solving with everything in one app under the same conditions with PJFNK took 22.51 s (compared to 21.40 s for just splitting it into a MultiApp) and there was a negligible difference in results.

Once all the systems are in place, I plan to scale up the number of nodes and timesteps to what is required for my problem and maybe the performance difference will matter more.

Maybe you are right that it may not be worth giving up full coupling for a ~20% improvement in performance.
